### PR TITLE
feat: gym wrapper for sebulba 

### PR DESCRIPTION
**Pull request by Louay-Ben-nessir**
## What?
Implemented a wrapper for gym environments to facilitate their use in sebulba.
## Why?
Integrate Sebulba's architecture due to its effectiveness in scenarios involving non-jitted/non-jax environments.
## How?
Add a wrapper for gym envs, the corresponding `create_env` function, and the sebulba yaml file.

### CODE REVIEW

File: mava/configs/arch/sebulba.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# --- Sebulba config ---
+arch_name: "sebulba"
```
Comment (OmaymaMahjoub) [inline mava/configs/arch/sebulba.yaml]: For unification, either remove arch_name or update anakin.yaml with arch_name

File: mava/configs/env/gym.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# ---Environment Configs---
+
+scenario: rware:rware-tiny-2ag-v1 # [tiny-2ag, tiny-4ag, tiny-4ag-easy, small-4ag]
```
Comment (OmaymaMahjoub) [inline mava/configs/env/gym.yaml]: didn't you make lbf works as well so maybe try to make either seperate files one for rware and one for lbf or merge them in one in a smart way

File: mava/utils/make_env.py [hunk 3]
```diff
@@ -14,8 +14,13 @@
 
 from typing import Tuple
 
+import gym
+import gym.vector
+import gym.wrappers
+import gym.wrappers.compatibility
 import jaxmarl
 import jumanji
+import lbforaging  # noqa: F401 used implicitly
```
Comment (OmaymaMahjoub) [inline mava/utils/make_env.py]: is there a reason for importing it without using?

File: mava/utils/make_env.py [hunk 4]
```diff
@@ -199,6 +211,43 @@ def make_gigastep_env(
...
+    """
+    base_env_name = config.env.env_name
+    wrapper = _gym_registry[base_env_name]
+
+    def create_gym_env(
+        config: DictConfig, add_global_state: bool = False
+    ) -> Environment:  # todo: add the RecordEpisodeMetrics for gym.
+        env = gym.make(config.env.scenario)
+        wrapped_env = wrapper(env, config.env.use_individual_rewards, add_global_state)
+        if not config.env.implicit_agent_id:
```
Comment (OmaymaMahjoub) [inline mava/utils/make_env.py]:
```suggestion
        if config.system.add_agent_id & (~config.env.implicit_agent_id):
```

File: mava/wrappers/gym.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    def step(self, action: list) -> Tuple[np.ndarray, float, bool, bool, Dict]:
+        """Step the environment."""
+        obs, reward, terminated, truncated, info = self.env.step(action)
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, reward, terminated, truncated, info
+
+
+# Copied form https://github.com/openai/gym/blob/master/gym/vector/async_vector_env.py
+# Modified to work with multiple agents
+def _multiagent_worker_shared_memory(  # noqa: CCR001
```
Comment (OmaymaMahjoub) [inline mava/wrappers/gym.py]: may I ask why we need this, does not async automatically handling memory and autoresetting?

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Thanks Louay! I just added some minor comments :)

File: requirements/requirements.txt [hunk 6]
```diff
@@ -9,6 +9,7 @@ jax
 jaxlib
 jaxmarl
 jumanji @ git+https://github.com/sash-a/jumanji
+lbforaging @ git+https://github.com/Louay-Ben-nessir/lb-foraging.git
```
Comment (sash-a) [inline requirements/requirements.txt]: Is there a reason we aren't using the original?

File: mava/configs/arch/anakin.yaml [hunk 7]
```diff
@@ -1,4 +1,5 @@
 # --- Anakin config ---
+arch_name: "Anakin"
```
Comment (sash-a) [inline mava/configs/arch/anakin.yaml]: seeing as we're keeping the name can we call it `architecture_name` and make it all lowercase

File: mava/configs/env/scenario/gym-10x10-3p-3f.yaml
Comment (sash-a) [inline mava/configs/env/scenario/gym-10x10-3p-3f.yaml:1]: can we name these files `gym-lfb-...` and while we're here can we prefix the lbf scenarios with `lbf-...` just so it's easier to navigate :pray: 

(Also will need to change things in the corresponding `lbf.yaml` file

But I'm confused why we even need this, isn't lbf registered with gym? Can't you just do `gym.make(10x10-3p-3f)`?

File: mava/configs/arch/anakin.yaml (hunk 7 above)
Comment (RuanJohn) [inline mava/configs/arch/anakin.yaml]:
```suggestion
arch_name: anakin
```

File: mava/utils/make_env.py [hunk 8]
```diff
@@ -14,8 +14,14 @@
 
 from typing import Tuple
 
+import gym
```
Comment (sash-a) [inline mava/utils/make_env.py]: We should be using gymnasium rather, gym is no longer supported

File: mava/configs/arch/sebulba.yaml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# --- Sebulba config ---
+arch_name: "Sebulba"
```
Comment (RuanJohn) [inline mava/configs/arch/sebulba.yaml]:
```suggestion
arch_name: sebulba
```

File: mava/configs/env/gym.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+# Weather or not to sum the returned rewards over all of the agents.
+use_shared_rewards: True
+
+kwargs:
+  {}
+
+# Possible scenarios:
+# RobotWarehouse : [tiny-2ag, tiny-4ag, tiny-4ag-easy, small-4ag]
+# LevelBasedForaging : [2s-8x8-2p-2f-coop, 8x8-2p-2f-coop, 2s-10x10-3p-3f, 10x10-3p-3f, 15x15-3p-5f, 15x15-4p-3f, 15x15-4p-5f]
```
Comment (RuanJohn) [inline mava/configs/env/gym.yaml]: Please use the write names for the config files here. I think the correct ones all how `gym-` in the names.

File: mava/wrappers/gym.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+        return agents_view, reward, terminated, truncated, info
+
+    def get_actions_mask(self, info: Dict) -> NDArray:
+        if "action_mask" in info:
+            return np.array(info["action_mask"])
+        return np.ones((self.num_agents, self.num_actions), dtype=np.float32)
+
+
+class GymRecordEpisodeMetrics(gym.Wrapper):
```
Comment (sash-a) [inline mava/wrappers/gym.py]: This already exists, can we rather use [this one](https://github.com/Farama-Foundation/Gymnasium/blob/main/gymnasium/wrappers/common.py#L434)?

File: mava/wrappers/gym.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+warnings.filterwarnings("ignore", module="gym.utils.passive_env_checker")
+
+
+class GymRwareWrapper(gym.Wrapper):
+    """Wrapper for rware gym environments."""
+
+    def __init__(
+        self,
+        env: gym.Env,
+        use_shared_rewards: bool = False,
```
Comment (RuanJohn) [inline mava/wrappers/gym.py]: Since we mostly focus on cooperative MARL, I think we can just have `use_shared_rewards` as True by default.

```suggestion
        use_shared_rewards: bool = True,
```

File: mava/wrappers/gym.py [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+        add_global_state: bool = False,
+    ):
+        """Initialize the gym wrapper
```
Comment (RuanJohn) [inline mava/wrappers/gym.py]:
```suggestion
        """Initialise the gym wrapper
```

File: mava/wrappers/gym.py [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+            add_global_state (bool, optional) : Create global observations. Defaults to False.
+        """
+        super().__init__(env)
+        self._env = env  # not having _env leaded tp self.env getting replaced --> circular called
+        self.use_shared_rewards = use_shared_rewards
+        self.add_global_state = add_global_state  # todo : add the global observations
+        self.num_agents = len(self._env.action_space)
+        self.num_actions = self._env.action_space[
+            0
+        ].n  # todo: all the agents must have the same num_actions, add assertion?
```
Comment (RuanJohn) [inline mava/wrappers/gym.py]: Did you ever implement that assertion? Otherwise maybe a warning saying that in LBF with assume agents are homogeneous.

File: mava/wrappers/gym.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+        Args:
+            env (gym.env): gym env instance.
+            use_shared_rewards (bool, optional): Use individual or shared rewards.
+            Defaults to False.
+            add_global_state (bool, optional) : Create global observations. Defaults to False.
+        """
+        super().__init__(env)
+        self._env = env  # not having _env leaded tp self.env getting replaced --> circular called
+        self.use_shared_rewards = use_shared_rewards
+        self.add_global_state = add_global_state  # todo : add the global observations
```
Comment (RuanJohn) [inline mava/wrappers/gym.py]: Did you add the global obs?

File: requirements/requirements.txt (hunk 6 above)
Comment (RuanJohn) [inline requirements/requirements.txt]: Just a general question, why are we depending on forks for the environments?

File: mava/wrappers/gym.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+        info = {"actions_mask": self.get_actions_mask(info)}
+
+        if self.use_shared_rewards:
+            reward = np.array([np.array(reward).sum()] * self.num_agents)
+        else:
+            reward = np.array(reward)
+
+        truncated = [truncated] * self.num_agents
+        terminated = [terminated] * self.num_agents
```
Comment (sash-a) [inline mava/wrappers/gym.py]:
```suggestion
        truncated = np.repeat(truncated, self.num_agents)
        terminated = np.repeat(terminated, self.num_agents)
```

File: mava/wrappers/__init__.py [hunk 17]
```diff
@@ -15,6 +15,13 @@
 from mava.wrappers.auto_reset_wrapper import AutoResetWrapper
 from mava.wrappers.episode_metrics import RecordEpisodeMetrics
 from mava.wrappers.gigastep import GigastepWrapper
+from mava.wrappers.gym import (
+    GymAgentIDWrapper,
+    GymLBFWrapper,
+    GymRecordEpisodeMetrics,
+    GymRwareWrapper,
+    _multiagent_worker_shared_memory,
```
Comment (sash-a) [inline mava/wrappers/__init__.py]: why export this if it's a private method?

File: mava/wrappers/gym.py (hunk 5 above)
Comment (sash-a) [reply mava/wrappers/gym.py]: I really don't like that we need to maintain this. Does pufferlib's wrapper not solve the multi agent/single agent issue?

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you for all the work @Louay-Ben-nessir!

Some minor comments and questions from me. But also a more general one, by default we have an env config that points to a scenario config. Do you think it'll be able to mirror this? For example, instead of having only an `mava/configs/env/gym.yaml` file do you think we could have `mava/configs/env/rware_gym.yaml` and `mava/configs/env/lbf_gym.yaml`?

File: mava/wrappers/gym.py (hunk 5 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py]: Yes, but i needed to change this line https://github.com/instadeepai/Mava/blob/133a25060151ccd99b0f0fe1a73af48310dbbbff/mava/wrappers/gym.py#L271
since originally it was:
```python3
if terminated or truncated:
```
leading to the environments resetting at each step when the terminated / truncated flags are returned as arrays.

File: mava/wrappers/gym.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+            observation_space.low[0],
+            observation_space.high[0],
+            observation_space.dtype,
+            observation_space.shape,
+        )
+        _new_obs_shape = (_obs_shape[0] + self.env.num_agents,)
+        _observation_boxs = [
+            spaces.Box(low=_obs_low, high=_obs_high, shape=_new_obs_shape, dtype=_obs_dtype)
+        ] * self.env.num_agents
+        self.observation_space = spaces.Tuple(_observation_boxs)
```
Comment (sash-a) [inline mava/wrappers/gym.py]: This will currently only work if the inner spaces are boxes. Can we make this more general?

File: mava/wrappers/gym.py [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+import numpy as np
+from gym import spaces
+from gym.vector.utils import write_to_shared_memory
+from numpy.typing import NDArray
+
+# Filter out the warnings
+warnings.filterwarnings("ignore", module="gym.utils.passive_env_checker")
+
+
+class GymRwareWrapper(gym.Wrapper):
```
Comment (sash-a) [inline mava/wrappers/gym.py]: I think it would make a lot of sense to have a base class here as we do for jumanji and jaxmarl and just have rware and lbf extend that and implement the action mask and global obs methods

File: mava/wrappers/gym.py [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        if seed is not None:
+            self.env.seed(seed)
+
+        agents_view, info = self._env.reset()
+
+        info = {"actions_mask": self.get_actions_mask(info)}
+
+        return np.array(agents_view), info
+
+    def step(self, actions: NDArray) -> Tuple:  # Vect auto rest
```
Comment (sash-a) [inline mava/wrappers/gym.py]: Not sure what this comment means

File: mava/wrappers/gym.py [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+    ):
+        """Initialize the gym wrapper
+        Args:
+            env (gym.env): gym env instance.
+            use_shared_rewards (bool, optional): Use individual or shared rewards.
+            Defaults to False.
+            add_global_state (bool, optional) : Create global observations. Defaults to False.
+        """
+        super().__init__(env)
+        self._env = env  # not having _env leaded tp self.env getting replaced --> circular called
```
Comment (sash-a) [inline mava/wrappers/gym.py]: This is strange, sure `super().__init__` should set this for you?

File: mava/wrappers/gym.py (hunk 14 above)
Comment (sash-a) [inline mava/wrappers/gym.py]: Agree with both of these todo's

File: mava/wrappers/gym.py [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+        super().__init__(env)
+        self._env = env
+        self.use_shared_rewards = use_shared_rewards
+        self.add_global_state = add_global_state
+        self.num_agents = len(self._env.action_space)
+        self.num_actions = self._env.action_space[0].n
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[np.ndarray, Dict]:
```
Comment (sash-a) [inline mava/wrappers/gym.py]:
```suggestion
    ) -> Tuple[NDArray, Dict]:
```

File: mava/wrappers/gym.py [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+
+        agents_view, info = self._env.reset()
+
+        info = {"actions_mask": self.get_actions_mask(info)}
+        if self.add_global_state:
+            info["global_obs"] = self.get_global_obs(agents_view)
+
+        return np.array(agents_view), info
+
+    def step(self, actions: NDArray) -> Tuple:
```
Comment (sash-a) [inline mava/wrappers/gym.py]:
```suggestion
    def step(self, actions: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray, Dict]:
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looking pretty good @Louay-Ben-nessir!

File: mava/wrappers/gym.py (hunk 15 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py]: just added them.

File: requirements/requirements.txt (hunk 6 above)
Comment (Louay-Ben-nessir) [reply requirements/requirements.txt]: The original implementation stalls when using newer versions of gym https://github.com/semitable/lb-foraging/issues/20

File: mava/wrappers/gym.py (hunk 11 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py]: It uses 
`if terminated or truncated:`
which leads to resetting every step if terminated/truncated are returned per agent

File: mava/configs/env/scenario/gym-10x10-3p-3f.yaml
Comment (Louay-Ben-nessir) [reply mava/configs/env/scenario/gym-10x10-3p-3f.yaml:1]: I changed it to use the config files for 2 reasons:
- More consistent with how the other envs work
- More flexibility as the registered envs might not contain the desired configs

Should allow for the use of both?

File: mava/wrappers/gym.py (hunk 20 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py]: I wrote some comments down while debugging and forgot to remove them 😅

File: mava/wrappers/gym.py [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+        add_global_state: bool = False,
+    ):
+        """Initialise the gym wrapper
+        Args:
+            env (gymnasium.env): gymnasium env instance.
+            use_shared_rewards (bool, optional): Use individual or shared rewards.
+            Defaults to False.
+            add_global_state (bool, optional) : Create global observations. Defaults to False.
+        """
+        super().__init__(env, use_shared_rewards, add_global_state)
```
Comment (sash-a) [inline mava/wrappers/gym.py]: You don't need the init if it's just calling super

```suggestion
```

File: mava/wrappers/gym.py [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+import numpy as np
+from gymnasium import spaces
+from gymnasium.vector.utils import write_to_shared_memory
+from numpy.typing import NDArray
+
+# Filter out the warnings
+warnings.filterwarnings("ignore", module="gymnasium.utils.passive_env_checker")
+
+
+class GymWrapper(gymnasium.Wrapper):
```
Comment (sash-a) [inline mava/wrappers/gym.py:32]: This is so much better! Nicely done :fire:

File: mava/wrappers/gym.py [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+        self._env = env
+        self.use_shared_rewards = use_shared_rewards
+        self.add_global_state = add_global_state
+        self.num_agents = len(self._env.action_space)
+        self.num_actions = self._env.action_space[0].n
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[NDArray, Dict]:
+
```
Comment (sash-a) [inline mava/wrappers/gym.py]: nit
```suggestion
    ) -> Tuple[NDArray, Dict]:
```

File: mava/wrappers/gym.py [hunk 27]
```diff
@@ -0,0 +1,299 @@
...
+        agents_view, info = self._env.reset()
+
+        info = {"actions_mask": self.get_actions_mask(info)}
+        if self.add_global_state:
+            info["global_obs"] = self.get_global_obs(agents_view)
+
+        return np.array(agents_view), info
+
+    def step(self, actions: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray, Dict]:
+
```
Comment (sash-a) [inline mava/wrappers/gym.py]: nit
```suggestion
    def step(self, actions: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray, Dict]:
```

File: mava/wrappers/gym.py [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+        Args:
+            env (gymnasium.env): gymnasium env instance.
+            use_shared_rewards (bool, optional): Use individual or shared rewards.
+            Defaults to False.
+            add_global_state (bool, optional) : Create global observations. Defaults to False.
+        """
+        super().__init__(env, use_shared_rewards, add_global_state)
+
+    def step(self, actions: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray, Dict]:
+
```
Comment (sash-a) [inline mava/wrappers/gym.py]: nit
```suggestion
    def step(self, actions: NDArray) -> Tuple[NDArray, NDArray, NDArray, NDArray, Dict]:
```

File: mava/wrappers/gym.py [hunk 29]
```diff
@@ -0,0 +1,299 @@
...
+        super().__init__(env)
+        self._env = env
+        self.running_count_episode_return = 0.0
+        self.running_count_episode_length = 0.0
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[NDArray, Dict]:
+
+        # Reset the env
```
Comment (sash-a) [inline mava/wrappers/gym.py]: nit
```suggestion
    ) -> Tuple[NDArray, Dict]:
```

File: mava/wrappers/gym.py [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+from gymnasium import spaces
+from gymnasium.vector.utils import write_to_shared_memory
+from numpy.typing import NDArray
+
+# Filter out the warnings
+warnings.filterwarnings("ignore", module="gymnasium.utils.passive_env_checker")
+
+
+class GymWrapper(gymnasium.Wrapper):
+    """Wrapper for gym environments."""
```
Comment (sash-a) [inline mava/wrappers/gym.py]: This line is likely too long, but if you can add in this comment so we know it works for rware

```suggestion
    """Base wrapper for multi-agent gym environments.
    
    This wrapper works out of the box for RobotWarehouse. See `GymLBFWrapper` for how it can be modified to work for other environments.
    """
```

File: mava/wrappers/gym.py [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+            return np.array(info["action_mask"])
+        return np.ones((self.num_agents, self.num_actions), dtype=np.float32)
+
+    def get_global_obs(self, obs: NDArray) -> NDArray:
+        global_obs = np.concatenate(obs, axis=0)
+        return np.tile(global_obs, (self.num_agents, 1))
+
+
+class GymLBFWrapper(GymWrapper):
+    """Wrapper for LBF gym environments"""
```
Comment (sash-a) [inline mava/wrappers/gym.py]:
```suggestion
    """Wrapper for the gym level based foraging environment."""
```

File: mava/wrappers/gym.py [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+        if "won_episode" in info:
+            metrics["won_episode"] = info["won_episode"]
+
+        info["metrics"] = metrics
+
+        return agents_view, info
+
+    def step(self, actions: NDArray) -> Tuple:
+
+        # Step the env
```
Comment (sash-a) [inline mava/wrappers/gym.py]: nit
```suggestion
    def step(self, actions: NDArray) -> Tuple:
```

File: mava/wrappers/gym.py [hunk 33]
```diff
@@ -0,0 +1,299 @@
...
+        _new_obs_shape = (_obs_shape[0] + self.env.num_agents,)
+        _observation_boxs = [
+            spaces.Box(low=_obs_low, high=_obs_high, shape=_new_obs_shape, dtype=_obs_dtype)
+        ] * self.env.num_agents
+        self.observation_space = spaces.Tuple(_observation_boxs)
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[NDArray, Dict]:
+        """Reset the environment."""
```
Comment (sash-a) [inline mava/wrappers/gym.py:171]:
```suggestion
```

File: mava/wrappers/gym.py [hunk 34]
```diff
@@ -0,0 +1,299 @@
...
+        obs, info = self.env.reset(seed, options)
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, info
+
+    def step(self, action: list) -> Tuple[NDArray, float, bool, bool, Dict]:
+        """Step the environment."""
```
Comment (sash-a) [inline mava/wrappers/gym.py:177]:
```suggestion
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looks really great now @Louay-Ben-nessir! Lot's of very small things, but then I think we're good to go!

File: mava/configs/env/scenario/gym-10x10-3p-3f.yaml
Comment (sash-a) [reply mava/configs/env/scenario/gym-10x10-3p-3f.yaml:1]: I see what you've done, I'm happy with how it works :smile:

File: mava/wrappers/gym.py (hunk 31 above)
Comment (WiemKhlifi) [reply mava/wrappers/gym.py]: `Level-Based Foraging` is  much better 🙏

File: mava/configs/default_ff_ippo.yaml [hunk 35]
```diff
@@ -3,5 +3,5 @@ defaults:
   - arch: anakin
   - system: ppo/ff_ippo
   - network: mlp
-  - env: rware
+  - env: rware_gym
```
Comment (WiemKhlifi) [inline mava/configs/default_ff_ippo.yaml]: Since you created another file for sebulba ff_ippo in this [PR,](https://github.com/instadeepai/Mava/pull/1088) could you please revert this change so it won't be merged with the other changes.

File: mava/utils/make_env.py [hunk 36]
```diff
@@ -199,6 +215,41 @@ def make_gigastep_env(
...
+    Args:
+        config (Dict): The configuration of the environment.
+        num_env (int) : The number of parallel envs to create.
+        add_global_state (bool): Whether to add the global state to the observation. Default False.
+
+    Returns:
+        Async environments.
+    """
+    base_env_name = config.env.scenario.name
+    env_maker, wrapper = _gym_registry[base_env_name]
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py]: `base_env_name ` is used once you can skip creating a variable for it:

```suggestion
    env_maker, wrapper = _gym_registry[config.env.scenario.name]
```
`

File: mava/utils/make_env.py [hunk 37]
```diff
@@ -199,6 +215,41 @@ def make_gigastep_env(
...
+
+    envs = gymnasium.vector.AsyncVectorEnv(
+        [lambda: create_gym_env(config, add_global_state) for _ in range(num_env)],
+        worker=async_multiagent_worker,
+    )
+
+    return envs
+
+
 def make(config: DictConfig, add_global_state: bool = False) -> Tuple[Environment, Environment]:
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py:252]: Could add here as well the if statement for gym envs in the `make()` function and avoid importing the `make_gym_env` directly in the system files to keep everything homogenous🙏

File: requirements/requirements.txt [hunk 38]
```diff
@@ -3,12 +3,14 @@ distrax @ git+https://github.com/google-deepmind/distrax  # distrax release does
...
 flax
 gigastep @ git+https://github.com/mlech26l/gigastep
+gymnasium
 hydra-core==1.3.2
 id-marl-eval @ git+https://github.com/instadeepai/marl-eval
 jax
 jaxlib
 jaxmarl
 jumanji @ git+https://github.com/sash-a/jumanji
+lbforaging @ git+https://github.com/LukasSchaefer/lb-foraging.git@gymnasium_integration
```
Comment (sash-a) [inline requirements/requirements.txt]:
```suggestion
lbforaging @ git+https://github.com/LukasSchaefer/lb-foraging.git@gymnasium_integration  # fixes: https://github.com/semitable/lb-foraging/issues/20
```

File: requirements/requirements.txt [hunk 39]
```diff
@@ -17,6 +19,7 @@ numpy
 omegaconf
 optax
 protobuf~=3.20
+rware @ git+https://github.com/RuanJohn/robotic-warehouse.git
```
Comment (sash-a) [inline requirements/requirements.txt]:
```suggestion
rware @ git+https://github.com/RuanJohn/robotic-warehouse.git  # compatibility with latest gym
```

File: mava/wrappers/gym.py [hunk 40]
```diff
@@ -0,0 +1,279 @@
...
+            )
+        elif isinstance(space, gymnasium.spaces.Tuple):
+            return gymnasium.spaces.Tuple(self.modify_space(s) for s in space)
+        else:
+            raise ValueError(f"Space {type(space)} is not currently supported.")
+
+
+# Copied form Gymnasium/blob/main/gymnasium/vector/async_vector_env.py
+# Modified to work with multiple agents
+def async_multiagent_worker(  # noqa CCR001
```
Comment (sash-a) [inline mava/wrappers/gym.py:196]: Are we keeping this in this PR and changing it later? If so can you just add a comment onto the lines that needed changing

**Review by sash-a — COMMENTED** [review]
Very minor things, but I think we're ready to go :bullettrain_side:

File: mava/wrappers/gym.py (hunk 14 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py]: There's a built-in check in gym's async workers. I removed the comment.

File: mava/wrappers/gym.py (hunk 40 above)
Comment (Louay-Ben-nessir) [reply mava/wrappers/gym.py:196]: The only difference is making the autotest able to handle arrays, I added a comment on top of the modified line.

**Review by sash-a — APPROVED** [review]
:1st_place_medal:

File: mava/utils/make_env.py [hunk 41]
```diff
@@ -199,6 +215,41 @@ def make_gigastep_env(
...
+        add_global_state (bool): Whether to add the global state to the observation. Default False.
+
+    Returns:
+        Async environments.
+    """
+    base_env_name = config.env.scenario.name
+    env_maker, wrapper = _gym_registry[base_env_name]
+
+    def create_gym_env(config: DictConfig, add_global_state: bool = False) -> Environment:
+        env = env_maker(**config.env.scenario.task_config)
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py:237]: Could you please set the `max_episode_steps=100` for all lbf as default and 500 for rware 🙏

File: mava/utils/make_env.py [hunk 42]
```diff
@@ -32,20 +36,27 @@
 from jumanji.environments.routing.robot_warehouse.generator import (
     RandomGenerator as RwareRandomGenerator,
 )
+from lbforaging.foraging import environment as gym_lbf
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py]: Does LBF work for you? I tested with the branch used for the ff_ippo PR and got this error `TypeError: 'module' object is not callable`. This should fix it:
```suggestion
from lbforaging.foraging import ForagingEnv as gym_lbf
```

And please fix the arguments naming in all configs, ie `num_agents -> players ` 
You'll find more info on the right naming [here](https://github.com/LukasSchaefer/lb-foraging/blob/master/lbforaging/__init__.py 🙌

**Review by WiemKhlifi — CHANGES_REQUESTED** [review]
Great work @Louay-Ben-nessir 🔥 Since most of the code is reviewed, I have a few minor comments and suggested fixes for some errors related to LBF 🙌

File: mava/utils/make_env.py (hunk 42 above)
Comment (Louay-Ben-nessir) [reply mava/utils/make_env.py]: I forgot to re-test lbf after the config file changes, everything should be working as intended now!

File: mava/utils/make_env.py (hunk 37 above)
Comment (Louay-Ben-nessir) [reply mava/utils/make_env.py:252]: This change is a bit tricky since the number of created envs varies depending on whether they are training, evaluation or absolute metric envs. 
Should I add the `num_envs` as a parameter to the `make()` function? it will only be used by the gym envs 
or should I make it return a function that creates the envs?

**Review by OmaymaMahjoub — APPROVED** [review]
Thanks @Louay-Ben-nessir :fire:

File: mava/utils/make_env.py (hunk 37 above)
Comment (WiemKhlifi) [reply mava/utils/make_env.py:252]: As discussed offline,  please just leave it as is it's confusing to see other Jax `make_env ` having this parameter 🙏
